openage/convert/value_object/read/media/smx.pyx

```diff
@@ -59,6 +57,415 @@ cdef public dict LAYER_TYPES = {
     1: SMXLayerType.SHADOW,
     2: SMXLayerType.OUTLINE,
 }
+cdef class SMXMainLayer8to5(SMXLayer):
+    def __init__(self, layer_header, data):
+        super().__init__(layer_header, data)
+        # process cmd table
+        for i in range(self.row_count):
+            cmd_offset, color_offset, chunk_pos, row_data = create_color_row(
+                self, i)
+            self.cmd_offset = cmd_offset
+            self.color_offset = color_offset
+            self.chunk_pos = chunk_pos
+
+            self.pcolor.push_back(row_data)
+
+cdef class SMXMainLayer4plus1(SMXLayer):
+    def __init__(self, layer_header, data):
+        super().__init__(layer_header, data)
+        # process cmd table
+        for i in range(self.row_count):
+            cmd_offset, color_offset, chunk_pos, row_data = create_color_row(
+                self, i)
+            self.cmd_offset = cmd_offset
+            self.color_offset = color_offset
+            self.chunk_pos = chunk_pos
+
+            self.pcolor.push_back(row_data)
+
+cdef class SMXOutlineLayer(SMXLayer):
+    def __init__(self, layer_header, data):
+        super().__init__(layer_header, data)
+        # process cmd table
+        for i in range(self.row_count):
+            cmd_offset, color_offset, chunk_pos, row_data = create_color_row(
+                self, i)
+            self.cmd_offset = cmd_offset
+            self.color_offset = color_offset
+            self.chunk_pos = chunk_pos
+
+            self.pcolor.push_back(row_data)
+
+cdef class SMXShadowLayer(SMXLayer):
+    def __init__(self, layer_header, data):
+        super().__init__(layer_header, data)
+        # process cmd table
+        for i in range(self.row_count):
+            cmd_offset, color_offset, chunk_pos, row_data = create_color_row(
+                self, i)
+            self.cmd_offset = cmd_offset
+            self.color_offset = color_offset
+            self.chunk_pos = chunk_pos
+
+            self.pcolor.push_back(row_data)
+
+
+ctypedef fused SMXLayerVariant:
+    SMXMainLayer8to5
+    SMXMainLayer4plus1
+    SMXOutlineLayer
+    SMXShadowLayer
```

Comment:
The typedef should be moved to the top of the file next to the other declarations of structs/enums.